File: 1-cafe-story-backend-javaspring/src/main/java/com/cafestory/entity/Region.java
package com.cafestory.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.Data;

import java.time.LocalDateTime;
import java.util.UUID;

@Data
@Entity
@Table(name = "regions")
public class Region {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "region_id", updatable = false, nullable = false)
    private UUID regionId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "city_code")
    private RegionCity cityRef;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "province_code")
    private RegionProvince provinceRef;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ward_code")
    private RegionWard wardRef;

    @Column(name = "city")
    private String city;

    @Column(name = "province")
    private String province;

    @Column(name = "ward")
    private String ward;

    @Column(name = "area")
    private String area;

    @Column(name = "street")
    private String street;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    void prePersist() {
        createdAt = LocalDateTime.now();
        updatedAt = createdAt;
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: 3-cafe-story-reactnative-mobile/src/types/region.ts
export type RegionProvinceResponse = {
  name: string;
  provinceCode: string;
};

export type RegionCityResponse = {
  cityCode: string;
  name: string;
  provinceCode: string;
};

export type RegionWardResponse = {
  cityCode: string;
  name: string;
  provinceCode: string;
  wardCode: string;
};

export type RegionRequirement = "FULL_ADDRESS" | "BLOG_LOCATION";

export type RegionRequest = {
  city?: string;
  cityCode?: string;
  province?: string;
  provinceCode?: string;
  street?: string;
  ward?: string;
  wardCode?: string;
};

export type RegionResponse = {
  area: string | null;
  city: string | null;
  cityCode: string | null;
  province: string | null;
  provinceCode: string | null;
  regionId: string;
  street: string | null;
  ward: string | null;
  wardCode: string | null;
};
